MaxText/llama_or_mistral_ckpt.py

```diff
@@ -230,6 +242,174 @@ def permute_to_match_maxtext_rope(arr):
+
+        lora_B_k = np.reshape(lora_B_k, [lora_rank, base_num_query_heads, head_dim])
+
+        if self_attention_lora["key"]["lora_a.kernel"] is None:
+          self_attention_lora["key"]["lora_a.kernel"] = np.zeros(stack_shape + lora_A_k.shape, dtype=np.float16)
+          self_attention_lora["key"]["lora_b.kernel"] = np.zeros(stack_shape + lora_B_k.shape, dtype=np.float16)
+
+        self_attention_lora["key"]["lora_a.kernel"][layer_idx, ...] = lora_A_k  # pylint: disable=E1137
+        self_attention_lora["key"]["lora_b.kernel"][layer_idx, ...] = lora_B_k  # pylint: disable=E1137
+
+      if "v_proj" in target_module:
+        lora_A_v = lora_chkpt_vars[f"layers.{layer_idx}.attention.wv.lora_A.weights"].type(torch.float16).numpy().transpose()
+        lora_B_v = lora_chkpt_vars[f"layers.{layer_idx}.attention.wv.lora_B.weights"].type(torch.float16).numpy().transpose()
+
+        lora_B_v = np.reshape(lora_B_v, [lora_rank, base_num_query_heads, head_dim])
+
+        if self_attention_lora["value"]["lora_a.kernel"] is None:
+          self_attention_lora["value"]["lora_a.kernel"] = np.zeros(stack_shape + lora_A_v.shape, dtype=np.float16)
+          self_attention_lora["value"]["lora_b.kernel"] = np.zeros(stack_shape + lora_B_v.shape, dtype=np.float16)
+
+        self_attention_lora["value"]["lora_a.kernel"][layer_idx, ...] = lora_A_v  # pylint: disable=E1137
+        self_attention_lora["value"]["lora_b.kernel"][layer_idx, ...] = lora_B_v  # pylint: disable=E1137
+
+      if "o_proj" in target_module:
+        lora_A_o = lora_chkpt_vars[f"layers.{layer_idx}.attention.wo.lora_A.weights"].type(torch.float16).numpy()
+        lora_B_o = lora_chkpt_vars[f"layers.{layer_idx}.attention.wo.lora_B.weights"].type(torch.float16).numpy()
+
+        # This is for "out" matrix. So we don't transpose it above as well as here
+        # we have to reshape the lora_A_o instead of lora_B_o.
+        lora_A_o = np.reshape(lora_A_o, [lora_rank, base_num_query_heads, head_dim])
+
+        if self_attention_lora["out"]["lora_a.kernel"] is None:
+          self_attention_lora["out"]["lora_a.kernel"] = np.zeros(stack_shape + lora_A_o.shape, dtype=np.float16)
+          self_attention_lora["out"]["lora_b.kernel"] = np.zeros(stack_shape + lora_B_o.shape, dtype=np.float16)
+
+        self_attention_lora["out"]["lora_a.kernel"][layer_idx, ...] = lora_A_o  # pylint: disable=E1137
+        self_attention_lora["out"]["lora_b.kernel"][layer_idx, ...] = lora_B_o  # pylint: disable=E1137# pylint: disable=E1137
+
+  if self_attention_lora["query"]["lora_a.kernel"] is not None:
+    self_attention_lora["query"]["lora_a.kernel"] = np.transpose(
+        self_attention_lora["query"]["lora_a.kernel"], axes=(1, 0, 2)
+    )
+    self_attention_lora["query"]["lora_b.kernel"] = np.transpose(
+        self_attention_lora["query"]["lora_b.kernel"], axes=(1, 0, 2, 3)
+    )
```

Comment:
Same with these. Is a helper method possible here?

Reply:
I don't think this helps a lot. In fact it is making confusing because we need to be careful for the shapes. 
I would  prefer to delegate this refactor until this whole file is refactored.